helpers/osinfo.go

```diff
@@ -0,0 +1,270 @@
+	OS_VERSION_CODENAME
+	OS_BUILD_ID
+	OS_IMAGE_ID
+	OS_IMAGE_VERSION
+	OS_KERNEL_RELEASE // not part of default os-release, but we can use it here to facilitate things
+)
+
+type OSReleaseField uint32
+
+func (o OSReleaseField) String() string {
+	return OSReleaseFieldToString[o]
+}
+
+// StringToOSReleaseField is a map of os-release file fields
+var StringToOSReleaseField = map[string]OSReleaseField{
+	"NAME":             OS_NAME,
+	"ID":               OS_ID,
+	"ID_LIKE":          OS_ID_LIKE,
+	"PRETTY_NAME":      OS_PRETTY_NAME,
+	"VARIANT":          OS_VARIANT,
+	"VARIANT_ID":       OS_VARIANT_ID,
+	"VERSION":          OS_VERSION,
+	"VERSION_ID":       OS_VERSION_ID,
+	"VERSION_CODENAME": OS_VERSION_CODENAME,
+	"BUILD_ID":         OS_BUILD_ID,
+	"IMAGE_ID":         OS_IMAGE_ID,
+	"IMAGE_VERSION":    OS_IMAGE_VERSION,
+	"KERNEL_RELEASE":   OS_KERNEL_RELEASE,
+}
+
+// OSReleaseFieldToString is a map of os-release file fields
+var OSReleaseFieldToString = map[OSReleaseField]string{
+	OS_NAME:             "NAME",
+	OS_ID:               "ID",
+	OS_ID_LIKE:          "ID_LIKE",
+	OS_PRETTY_NAME:      "PRETTY_NAME",
+	OS_VARIANT:          "VARIANT",
+	OS_VARIANT_ID:       "VARIANT_ID",
+	OS_VERSION:          "VERSION",
+	OS_VERSION_ID:       "VERSION_ID",
+	OS_VERSION_CODENAME: "VERSION_CODENAME",
+	OS_BUILD_ID:         "BUILD_ID",
+	OS_IMAGE_ID:         "IMAGE_ID",
+	OS_IMAGE_VERSION:    "IMAGE_VERSION",
+	OS_KERNEL_RELEASE:   "KERNEL_RELEASE",
+}
+
+// CompareOSBaseKernelRelease will compare two given kernel version/release
+// strings and return -1, 0 or 1 if given version is less, equal or bigger,
+// respectively, than the given one
+//
+// Examples of $(uname -r):
+//
+// 5.11.0-31-generic (ubuntu)
+// 4.18.0-305.12.1.el8_4.x86_64 (alma)
+// 4.18.0-338.el8.x86_64 (stream8)
+// 4.18.0-305.7.1.el8_4.centos.x86_64 (centos)
+// 4.18.0-305.7.1.el8_4.centos.plus.x86_64 (centos + plus repo)
+// 5.13.13-arch1-1 (archlinux)
+//
+func CompareOSBaseKernelRelease(base, given string) int {
+	b := strings.Split(base, "-") // [base]-xxx
+	b = strings.Split(b[0], ".")  // [major][minor][patch]
+
+	g := strings.Split(given, "-")
+	g = strings.Split(g[0], ".")
+
+	for n := 0; n <= 2; n++ {
+		i, _ := strconv.Atoi(g[n])
+		j, _ := strconv.Atoi(b[n])
+
+		if i > j {
+			return 1 // given is bigger
+		} else if i < j {
+			return -1 // given is less
+		} else {
+			continue // equal
+		}
+	}
+
+	return 0 // equal
+}
+
+// UnameRelease gets the version string of the current running kernel
+func UnameRelease() (string, error) {
+	var uname syscall.Utsname
+	if err := syscall.Uname(&uname); err != nil {
+		return "", fmt.Errorf("could not get utsname")
+	}
+
+	var buf [65]byte
+	for i, b := range uname.Release {
+		buf[i] = byte(b)
+	}
+
+	ver := string(buf[:])
+	ver = strings.Trim(ver, "\x00")
+
+	return ver, nil
+}
+
+// OSBTFEnabled checks if kernel has embedded BTF vmlinux file
+func OSBTFEnabled() bool {
+	_, err := os.Stat("/sys/kernel/btf/vmlinux")
```

Comment:
Is this the only way of doing this? Does checking the kconfig make sense at all?

Reply:
Thats a good question actually @grantseltzer ... There might be other ways. I think that if you read the ELF file and it contains the .BTF section its half way done. You would still miss whether BTF is exported through sysfs, which you would have to check kconfig. Maybe reading kconfig and assuming BTF is enabled is better than trying to see if that file exists (just like any other kconfig feature. If not done by this PR, that is something we should do indeed.

Reply:
I'm adding a TODO now .. so OSBTFEnabled() gets a KernelConfig type as parameter and checks if BTF kernel config option is set for the running environment. Dont think we should in this PR as this is being coordinated with tracee and has enough changes already.